feat: switch to body.view (implement #285)

PiiXiieeS opened the pull request:
Switch from body.styled_view to body.view.
The change had a side effect on loosing the `<div>` with ID `Content` which was provided by `body.styled_view` so we have to build it via cheerio when the context is initiated.
Most of the test changes are triggered by this missing `<div>` on the initial html.

Resolves: #285

PiiXiieeS committed:
feat: switch to body.view (implement #285)

Switch from body.styled_view to body.view.
The change had a side effect on loosing the <div> with ID `Content`
which was aprovided by `body.styled_view` so we have to build it via
cheerio when the context is initiated.
Most of the test changes are triggered by this missing <div> on
the initial html.

Resolves: #285

MeStrak commented on src/context/context.service.ts:
```diff
@@ -95,7 +95,11 @@ export class ContextService {
   }
 
   setHtmlBody(body: string): void {
```
Worked for me but I noticed that it still wraps the loaded doc in `<html>` tags. A suggested modification below that would allow this function to be used by both the page-proxy and the new page api call so there's no need for two separate functions:

```suggestion
     setHtmlBody(
    body: string,
    loadAsDocument = false,
    divId = 'konviw-ontent',
  ): void {
    const $ = cheerio.load(body);
    // we wrap the body in a div with ID Content
    this.cheerioBody = cheerio.load(
      $('html').wrapInner(`<div id="${divId}">`).html(),
      null,
      loadAsDocument,
    );
  }
```
Explanation of the suggested new parameters (note that the default values are just proposals and there's no problem if you want to go with different defaults.

**`loadAsDocument`**: sets the cheerio `isDocument` property. If true it will load the page wrapped with `<html>`, if false it will load it as a fragment so it does not add `<html></html>` which is the desired behaviour for the page API (more info in the cheerio docs: https://cheerio.js.org/modules.html#load). I proposed the default as `false` as the method name gives the expectation that the body is loaded rather than an entire document.

**`divId`**: I think it would be useful to be able to customise this ID, especially as `Content` is quite generically used. For example if a site is embedding the page, there might be another div with `id=Content`, or if there is a need to embed two pages in one there would be duplicate div IDs. I proposed the default here as `konviw-content` to have a less generic ID, but I guess this might cause some issues in existing stylesheets so it's no issue if `Content` is used.

MeStrak commented on src/proxy-page/proxy-page.service.ts:
```diff
@@ -52,7 +52,7 @@ export class ProxyPageService {
     this.context.Init(spaceKey, pageId, theme, style);
     this.context.setTitle(results.title);
-    this.context.setHtmlBody(results.body.styled_view.value);
+    this.context.setHtmlBody(results.body.view.value);
```
Great!

It worked almost first time for me, and if we incorporate my other suggestion about the setHtmlBody function we can use the same code for loading page HTML whether it's for the page-proxy or page API.

MeStrak replied:
Just realised that the `initContext` function would also need updating as that's the one that calls `setHtmlBody` . In the new page API code I have currently changed that to a central function as it is duplicated between the page API page-proxy services but it would still be pretty simple to handle that logic.

PiiXiieeS committed:
fix: generic setHtmlBody for Page and API modules

Worked for me but I noticed that it still wraps the loaded doc in <html> tags. A suggested modification below that would allow this function to be used by both the page-proxy and the new page api call so there's no need for two separate functions:

Co-authored-by: MeStrak <31989238+MeStrak@users.noreply.github.com>

PiiXiieeS commented on src/context/context.service.ts:
```diff
@@ -94,8 +94,24 @@ export class ContextService {
+     setHtmlBody(
+    body: string,
+    loadAsDocument = false,
+    divId = 'konviw-ontent',
```
I would not make the `div Id` configurable because there are too much dependencies on the code and stylesheets on this one. So better to keep it by convention as #Content.

MeStrak replied:
Ok, makes sense and removes a bit of complexity in the implementation - an embedding app can put it inside another div and it will work fine.

PiiXiieeS commented on src/context/context.service.ts:
```diff
@@ -94,8 +94,24 @@ export class ContextService {
-    this.cheerioBody = cheerio.load(body);
+     setHtmlBody(
+    body: string,
+    loadAsDocument = false,
```
I will keep your suggestion while prefer to `loadAsDocument = true` by default to keep compatibility with former usage in renderPage and renderSlides.
However, it may be easier and obvious just to define two utilities to load the body in the context module:
- `setHtmlBody` as it was with `loadAsDocument = true` by default
- `setBody` with loadAsDocument = false` for the needed behaviour in the API

PiiXiieeS committed:
fix(context): setHtmlBody allowing fragmented mode

isDocument: boolean = true by default while it allows now to pass isDocument = false
to allow the parser to be switched to fragment mode.

PiiXiieeS commented on src/context/context.service.ts:
```diff
@@ -94,25 +94,15 @@ export class ContextService {
-    loadAsDocument = false,
-    divId = 'konviw-ontent',
-  ): void {
+  setHtmlBody(body: string, loadAsDocument = true): void {
```
I have kept your suggestion while set `loadAsDocument = true` by default to keep compatibility with former usage in `renderPage` and `renderSlides` and many tests.
However, it may be easier and less prone to errors just to define two utilities to load the body in the context module:
- `setHtmlBody` as it was with `loadAsDocument = true` by default
- `setBody` with `loadAsDocument = false` for the needed behaviour in the API
